src/test/crypto_tests.cpp

```diff
@@ -857,4 +859,88 @@ BOOST_AUTO_TEST_CASE(sha3_256_tests)
+    unsigned char tmp[32] = {i, 0};
+    return MuHash3072(tmp);
+}
+
+BOOST_AUTO_TEST_CASE(muhash_tests)
+{
+    unsigned char out[384];
+
+    for (int iter = 0; iter < 10; ++iter) {
+        unsigned char res[384];
+        int table[4];
+        for (int i = 0; i < 4; ++i) {
+            table[i] = g_insecure_rand_ctx.randbits(3);
+        }
+        for (int order = 0; order < 4; ++order) {
+            MuHash3072 acc;
+            for (int i = 0; i < 4; ++i) {
+                int t = table[i ^ order];
+                if (t & 4) {
+                    acc /= FromInt(t & 3);
+                } else {
+                    acc *= FromInt(t & 3);
+                }
+            }
+            acc.Finalize(out);
+            if (order == 0) {
+                memcpy(res, out, 384);
+            } else {
+                BOOST_CHECK(memcmp(res, out, 384) == 0);
+            }
+        }
+
+        MuHash3072 x = FromInt(g_insecure_rand_ctx.randbits(4)); // x=X
+        MuHash3072 y = FromInt(g_insecure_rand_ctx.randbits(4)); // x=X, y=Y
+        MuHash3072 z; // x=X, y=Y, z=1
+        z *= x; // x=X, y=Y, z=X
+        z *= y; // x=X, y=Y, z=X*Y
+        y *= x; // x=X, y=Y*X, z=X*Y
+        z /= y; // x=X, y=Y*X, z=1
+        z.Finalize(out);
+        for (int i = 0; i < 384; ++i) {
+            BOOST_CHECK_EQUAL(out[i], i == 0);
+        }
+    }
+
+    MuHash3072 acc = FromInt(0);
+    acc *= FromInt(1);
+    acc /= FromInt(2);
+    acc.Finalize(out);
+    uint256 x = (CHashWriter(SER_DISK, 0) << out).GetSHA256();
+    BOOST_CHECK(x == uint256S("a44e16d5e34d259b349af21c06e65d653915d2e208e4e03f389af750dc0bfdc3"));
+
+    // Test MuHash3072 serialization
+    MuHash3072 serchk = FromInt(1); serchk *= FromInt(2);
+    std::string ser_exp = "ad5d5a19c789b21f95f9c2a09f264f8aab4ef29ca30824f7330a8c3f4c61c25a6035bc27476ade3a38f2f2bc576bfbeabdf69641a91b99631ef904d37103eff744129842b1fd83158db6466ab4b752278f04add175a7a41ab8ea1305f66855068730dd3baf1ad29d6e13c7ba3350864a8f96e2e9e5c1ca65e11ab3e361184213a3fb395ac5c86efe3997c671cc5d46df3e1b2f00bf09120ae252a0e985b2452be64d5f626c0bc03747355d91ff2aa31ea4087797b91c80af5003d4b3a2ba5c34cd4a8785712897cd2331bac1b749e4fe9d17a5bb68395e2571d1759dceb5609dfae38ec6186e8c2eb22ffeb9bb8330083ee15d4b816e117e6bc1399c707b3e9d01ae74d2f589d6eb8e9e58c572088476e0444e3dbc05c95867acdb7c18be551e0eb2189d3c39dd86dbc844dc70ac33f9328be9e3b0d792bb359498f5fb385e2d8a8e3a5cde7e3e4e8ba123d3e1eda0b859d3a56e548dd6d23a03ce82e01493f96a67abd3109f9f4315c966f8cba311d71899c04a7fd9618ed028580ffcd21263";
+    CDataStream ss_chk(SER_DISK, PROTOCOL_VERSION);
+    ss_chk << serchk;
+    BOOST_CHECK_EQUAL(ser_exp, HexStr(ss_chk.str()));
+
+    // Test MuHash3072 deserialization
+    MuHash3072 deserchk;
+    ss_chk >> deserchk;
+    unsigned char out2[384];
+    serchk.Finalize(out);
+    deserchk.Finalize(out2);
+    BOOST_CHECK_EQUAL(HexStr(out), HexStr(out2));
+
+    // Test MuHash3072 overflow
+    class MuHashOverflow : public MuHash3072 {
```

Comment:
Again, this isn't actually overflowing the data type in MuHash. It's simply reducing a number greater than the group order into its least residue. Consider renaming this function/class.

Reply:
Since the overflow naming is consistent with secp256k1 I kept it but extended the comment for clarification.